File: java/core/src/main/java/network/aika/typedefs/SynapseDefinition.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.typedefs;

import network.aika.activations.*;
import network.aika.bindingsignal.BSType;
import network.aika.neurons.ConjunctiveSynapse;
import network.aika.neurons.DisjunctiveSynapse;
import network.aika.type.relations.Relation;
import network.aika.type.relations.RelationMany;
import network.aika.type.relations.RelationOne;
import network.aika.neurons.Neuron;
import network.aika.neurons.Synapse;
import network.aika.bindingsignal.Transition;
import network.aika.misc.direction.Direction;
import network.aika.type.Type;
import network.aika.type.TypeRegistry;
import network.aika.type.relations.RelationSelf;

import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

/**
 *
 * @author Lukas Molzberger
 */
public class SynapseDefinition extends Type {

    public enum SynapseSubType {
        CONJUNCTIVE,
        DISJUNCTIVE
    }

    public static final RelationSelf SELF = new RelationSelf(0, "SYN-SELF");

    public static final RelationOne INPUT = new RelationOne(1, "SYN-INPUT");
    public static final RelationOne OUTPUT = new RelationOne(2, "SYN-OUTPUT");
    public static final RelationMany LINK = new RelationMany(3, "SYN-LINK");

    public static final Relation[] RELATIONS = {SELF, INPUT, OUTPUT, LINK};

    static {
        LINK.setReversed(LinkDefinition.SYNAPSE);
        INPUT.setReversed(NeuronDefinition.OUTPUT);
        OUTPUT.setReversed(NeuronDefinition.INPUT);
    }

    SynapseSubType subType;

    private LinkDefinition link;

    private NeuronDefinition input;
    private NeuronDefinition output;

    private Transition[] transition;

    private Direction storedAt;

    private Boolean trainingAllowed;

    private SynapseDefinition instanceSynapseType;

    public SynapseDefinition(TypeRegistry registry, String name) {
        super(registry, name);
    }

    @Override
    public Relation[] getRelations() {
        return RELATIONS;
    }

    public Synapse instantiate() {
        switch (subType) {
            case CONJUNCTIVE:
                return new ConjunctiveSynapse(this);
            case DISJUNCTIVE:
                return new DisjunctiveSynapse(this);

            default:
                return null;
        }
    }

    public Synapse instantiate(Neuron input, Neuron output) {
        assert input.getType().isInstanceOf(getInput());
        assert output.getType().isInstanceOf(getOutput());

        switch (subType) {
            case CONJUNCTIVE:
                return new ConjunctiveSynapse(this, input, output);
            case DISJUNCTIVE:
                return new DisjunctiveSynapse(this, input, output);
            default:
                return null;
        }
    }

    public SynapseSubType getSubType() {
        return subType;
    }

    public SynapseDefinition setSubType(SynapseSubType subType) {
        this.subType = subType;
        return this;
    }

    public NeuronDefinition getInput() {
        return input != null ?
                input :
                getFromParent(p -> ((SynapseDefinition)p).getInput());
    }

    public SynapseDefinition setInput(NeuronDefinition input) {
        assert input != null;

        this.input = input;

        return this;
    }

    public NeuronDefinition getOutput() {
        return output != null ?
                output :
                getFromParent(p -> ((SynapseDefinition)p).getOutput());
    }

    public SynapseDefinition setOutput(NeuronDefinition outputDef) {
        assert outputDef != null;

        this.output = outputDef;

        return this;
    }

    public LinkDefinition getLink() {
        return link != null ?
                link :
                getFromParent(p -> ((SynapseDefinition)p).getLink());
    }

    public SynapseDefinition setLink(LinkDefinition link) {
        assert link != null;

        this.link = link;

        return this;
    }

    public boolean isIncomingLinkingCandidate(Set<BSType> BSTypes) {
        for (Transition t : getTransition()) {
            if (BSTypes.contains(t.to()))
                return true;
        }
        return false;
    }

    public boolean isOutgoingLinkingCandidate(Set<BSType> BSTypes) {
        for (Transition t : getTransition()) {
            if (BSTypes.contains(t.from()))
                return true;
        }
        return false;
    }

    public BSType mapTransitionForward(BSType bsType) {
        for (Transition t : getTransition()) {
            if(t.from() == bsType)
                return t.to();
        }
        return null;
    }

    public BSType mapTransitionBackward(BSType bsType) {
        for (Transition t : getTransition()) {
            if(t.to() == bsType)
                return t.from();
        }
        return null;
    }

    public Transition[] getTransition() {
        return transition != null ?
                transition :
                getFromParent(p -> ((SynapseDefinition)p).getTransition());
    }

    public SynapseDefinition setTransition(Transition... transition) {
        this.transition = transition;

        return this;
    }

    public Direction getStoredAt() {
        return storedAt != null ?
                storedAt :
                getFromParent(p -> ((SynapseDefinition)p).getStoredAt());
    }

    public SynapseDefinition setStoredAt(Direction storedAt) {
        this.storedAt = storedAt;

        return this;
    }

    public SynapseDefinition setTrainingAllowed(boolean trainingAllowed) {
        this.trainingAllowed = trainingAllowed;

        return this;
    }

    public SynapseDefinition getInstanceSynapseType() {
        return instanceSynapseType != null ?
                instanceSynapseType :
                getFromParent(p -> ((SynapseDefinition)p).getInstanceSynapseType());
    }

    public SynapseDefinition setInstanceSynapseType(SynapseDefinition instanceSynapseType) {
        this.instanceSynapseType = instanceSynapseType;

        return this;
    }

    @Override
    public String toString() {
        return super.toString() + " (" + getInput().getName() + " --> " + getOutput().getName() + ")";
    }
}

File: src/network/synapse_definition.cpp
#include "network/synapse_definition.h"

const RelationSelf SynapseDefinition::SELF = RelationSelf(0, "SELF");
const RelationOne SynapseDefinition::INPUT = RelationOne(1, "INPUT");
const RelationOne SynapseDefinition::OUTPUT = RelationOne(2, "OUTPUT");
const RelationMany SynapseDefinition::LINK = RelationMany(3, "LINK");
// Cannot store abstract class Relation in vector, need to use pointers instead
// const std::vector<Relation> SynapseDefinition::RELATIONS = {SELF, INPUT, OUTPUT, LINK};

// Static initializer to set up reverse relationships
class SynapseDefinitionInitializer {
public:
    SynapseDefinitionInitializer() {
        // Set up bidirectional relationships
        const_cast<RelationOne&>(SynapseDefinition::INPUT).setReversed(const_cast<RelationOne*>(&SynapseDefinition::OUTPUT));
        const_cast<RelationOne&>(SynapseDefinition::OUTPUT).setReversed(const_cast<RelationOne*>(&SynapseDefinition::INPUT));
        
        // SELF and LINK are their own reverse
        const_cast<RelationSelf&>(SynapseDefinition::SELF).setReversed(const_cast<RelationSelf*>(&SynapseDefinition::SELF));
        const_cast<RelationMany&>(SynapseDefinition::LINK).setReversed(const_cast<RelationMany*>(&SynapseDefinition::LINK));
    }
};

static SynapseDefinitionInitializer synapseDefInit;

SynapseDefinition::SynapseDefinition(TypeRegistry* registry, const std::string& name) : Type(registry, name), subType(SynapseSubType::CONJUNCTIVE), link(nullptr), input(nullptr), output(nullptr), storedAt(nullptr), trainingAllowed(false), instanceSynapseType(nullptr) {}

std::vector<Relation> SynapseDefinition::getRelations() const {
    // We can't return a vector of the abstract class Relation
    // For now, return an empty vector
    return std::vector<Relation>();
}

Synapse* SynapseDefinition::instantiate() {
    // Implementation for instantiating a Synapse
    return nullptr; // Placeholder
}

Synapse* SynapseDefinition::instantiate(Neuron* input, Neuron* output) {
    // Implementation for instantiating a Synapse with input and output
    return nullptr; // Placeholder
}

SynapseDefinition::SynapseSubType SynapseDefinition::getSubType() const {
    return subType;
}

SynapseDefinition* SynapseDefinition::setSubType(SynapseSubType subType) {
    this->subType = subType;
    return this;
}

NeuronType* SynapseDefinition::getInput() const {
    return input;
}

SynapseDefinition* SynapseDefinition::setInput(NeuronType* input) {
    this->input = input;
    return this;
}

NeuronType* SynapseDefinition::getOutput() const {
    return output;
}

SynapseDefinition* SynapseDefinition::setOutput(NeuronType* outputDef) {
    this->output = outputDef;
    return this;
}

LinkDefinition* SynapseDefinition::getLink() const {
    return link;
}

SynapseDefinition* SynapseDefinition::setLink(LinkDefinition* link) {
    this->link = link;
    return this;
}

bool SynapseDefinition::isIncomingLinkingCandidate(const std::set<BSType*>& BSTypes) const {
    // Implementation for checking incoming linking candidate
    return false; // Placeholder
}

bool SynapseDefinition::isOutgoingLinkingCandidate(const std::set<BSType*>& BSTypes) const {
    // Implementation for checking outgoing linking candidate
    return false; // Placeholder
}

BSType* SynapseDefinition::mapTransitionForward(BSType* bsType) const {
    // Implementation for mapping transition forward
    return nullptr; // Placeholder
}

BSType* SynapseDefinition::mapTransitionBackward(BSType* bsType) const {
    // Implementation for mapping transition backward
    return nullptr; // Placeholder
}

std::vector<Transition*> SynapseDefinition::getTransition() const {
    return transition;
}

SynapseDefinition* SynapseDefinition::setTransition(const std::vector<Transition*>& transition) {
    this->transition = transition;
    return this;
}

NetworkDirection* SynapseDefinition::getStoredAt() const {
    return storedAt;
}

SynapseDefinition* SynapseDefinition::setStoredAt(NetworkDirection* storedAt) {
    this->storedAt = storedAt;
    return this;
}

SynapseDefinition* SynapseDefinition::setTrainingAllowed(bool trainingAllowed) {
    this->trainingAllowed = trainingAllowed;
    return this;
}

SynapseDefinition* SynapseDefinition::getInstanceSynapseType() const {
    return instanceSynapseType;
}

SynapseDefinition* SynapseDefinition::setInstanceSynapseType(SynapseDefinition* instanceSynapseType) {
    this->instanceSynapseType = instanceSynapseType;
    return this;
}

std::string SynapseDefinition::toString() const {
    return "SynapseDefinition: " + name;
} 